File: application/backend-credit-card-enrollment/backend-c#/CreditCardEnrollment/Common/Command/CommandController.cs
using CreditCardEnrollment.Common.EventStore;
using CreditCardEnrollment.Common.Projection;

namespace CreditCardEnrollment.Common.Command;

public class CommandController {
    private readonly PostgresTransactionalEventStore _postgresTransactionalEventStore;
    private readonly MongoTransactionalProjectionOperator _mongoTransactionalProjectionOperator;

    public CommandController(
        PostgresTransactionalEventStore postgresTransactionalEventStore,
        MongoTransactionalProjectionOperator mongoTransactionalProjectionOperator) {
        _postgresTransactionalEventStore = postgresTransactionalEventStore;
        _mongoTransactionalProjectionOperator = mongoTransactionalProjectionOperator;
    }

    public void ProcessCommand(Command command, CommandHandler commandHandler) {
        try {
            _postgresTransactionalEventStore.BeginTransaction();
            _mongoTransactionalProjectionOperator.StartTransaction();
            commandHandler.HandleCommand(command);
            _postgresTransactionalEventStore.CommitTransaction();
            _mongoTransactionalProjectionOperator.CommitTransaction();

            _postgresTransactionalEventStore.AbortDanglingTransactionsAndReturnConnectionToPool();
            _mongoTransactionalProjectionOperator.AbortDanglingTransactionsAndReturnSessionToPool();
        } catch (Exception ex) {
            _postgresTransactionalEventStore.AbortDanglingTransactionsAndReturnConnectionToPool();
            _mongoTransactionalProjectionOperator.AbortDanglingTransactionsAndReturnSessionToPool();
            throw new Exception($"Failed to process command: {ex.Message}", ex);
        }
    }
}

File: application/backend-credit-card-enrollment/backend-typescript/src/common/command/CommandController.ts
import { PostgresTransactionalEventStore } from '../eventStore/PostgresTransactionalEventStore';
import { MongoTransactionalProjectionOperator } from '../projection/MongoTransactionalProjectionOperator';
import { log } from '../util/Logger';
import { Command } from './Command';
import { CommandHandler } from './CommandHandler';

export class CommandController {
    constructor(
        private readonly postgresTransactionalEventStore: PostgresTransactionalEventStore,
        private readonly mongoTransactionalProjectionOperator: MongoTransactionalProjectionOperator,
    ) {}

    protected async processCommand(command: Command, commandHandler: CommandHandler): Promise<void> {
        try {
            log.debug(`Starting to process command: ${command.constructor.name}`);
            await this.postgresTransactionalEventStore.beginTransaction();
            await this.mongoTransactionalProjectionOperator.startTransaction();
            await commandHandler.handleCommand(command);
            await this.postgresTransactionalEventStore.commitTransaction();
            await this.mongoTransactionalProjectionOperator.commitTransaction();

            await this.postgresTransactionalEventStore.abortDanglingTransactionsAndReturnConnectionToPool();
            await this.mongoTransactionalProjectionOperator.abortDanglingTransactionsAndReturnSessionToPool();
            log.debug(`Successfully processed command: ${command.constructor.name}`);
        } catch (error) {
            await this.postgresTransactionalEventStore.abortDanglingTransactionsAndReturnConnectionToPool();
            await this.mongoTransactionalProjectionOperator.abortDanglingTransactionsAndReturnSessionToPool();
            log.error(`Exception in ProcessCommand: ${error}`, error as Error);
            throw new Error(`Failed to process query: ${error}`);
        }
    }
}